NMS-13760: Split SNMP Property Extenders into multiple pages (#4086)
OpenNMS/opennms: docs/modules/operation/pages/performance-data-collection/snmp-property-extenders/regex.adoc

> == Regex Property Extender
> 
> The Regex property extender works similarly to the Index Split property extender, with the added capability of importing a string property from a source table.
> 
> === When to Use the Regex Property Extender

Reviewer: ```suggestion
= Regex Property Extender

The Regex property extender works similarly to the Index Split property extender, with the added capability of importing a string property from a source table.

== When to Use the Regex Property Extender
```

For example, the Cisco Airespace `bsnAPIfLoadParametersTable` is indexed using the tuple of `bsnAPDot3MacAdddress` and `bsnAPIfSlotId`, whereas the `bsnAPTable` is indexed on `bsnAPDot3MacAddress` alone.

[source, snmp-mib]
----
bsnAPIfLoadParametersEntry OBJECT-TYPE
    -- ...
    DESCRIPTION
        "An entry (conceptual row) in the Table.
        Entries in this MIB are indexed by
        bsnAPDot3MacAddress and bsnAPIfSlotId"
    INDEX           {
                        bsnAPDot3MacAddress,
                        bsnAPIfSlotId
                    } <1>

    -- ...

bsnAPEntry OBJECT-TYPE
    -- ...
    DESCRIPTION
        "An entry in the bsnAPTable."
    INDEX           { bsnAPDot3MacAddress } <2>
    -- ...
----
<1> `bsnAPDot3MacAddress` is the first component of the compound index for the entry type for `bsnAPIfLoadParametersTable`
<2> `bsnAPDot3MacAddress` is the sole index for the entry type for `bsnAPTable`

By extracting just the first index component and using the result as an index into the source MIB table, it's possible to import the human-sensible `bsnAPName` string property from the source MIB table.

> === Configuring the Regex Property Extender

Reviewer: ```suggestion
== Configuring the Regex Property Extender
```

The Regex property extender expects three parameters, all of which are required:

.Regex Property Extender Parameters
[options="header", cols="1,3"]
|===
| Name
| Description

| source-type

> | The name of the `resourceType` associated with the source MIB table

Reviewer 2: ```suggestion
| The name of the `resourceType` associated with the source MIB table.
```

| source-alias

> | The alias name of the string property to be imported from the source MIB table

Reviewer 2: ```suggestion
| The alias name of the string property to import from the source MIB table.
```

| index-pattern

> | A regular expression containing one matching group

Reviewer 2: ```suggestion
| A regular expression containing one matching group.
```

This example shows how to use the value of `bsnAPDot3MacAddress` as an index into the `bsnAPTable`.

[source, xml]
----
<resourceType name="bsnAPEntry" label="Cisco Wireless AP" resourceLabel="${bsnAPName} (index ${index})">
  <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy" />
  <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy" />
</resourceType>

<resourceType name="bsnAPIfLoadParametersEntry" label="Cisco Wireless AP Resources" resourceLabel="${bsnAPName} (index ${index})">
  <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy" />
  <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy" />
</resourceType>

<groups>
  <group name="bsnAPTable" ifType="all">
    <mibObj oid=".1.3.6.1.4.1.14179.2.2.1.1.3" instance="bsnAPEntry" alias="bsnAPName" type="string" /> <1>
  </group>

  <group name="bsnAPIfLoadParametersTable" ifType="all">
    <mibObj oid=".1.3.6.1.4.1.14179.2.2.13.1.4" instance="bsnAPIfLoadParametersEntry" alias="bsnAPIfLoadNumOfCli" type="integer" />

>     <property instance="bsnAPIfLoadParametersEntry" alias="bsnAPName" class-name="org.opennms.netmgt.collectd.RegexPropertyExtender"> <2>

Reviewer: ```suggestion
    <property instance="bsnAPIfLoadParametersEntry" alias="bsnAPName" class-name="org.opennms.netmgt.collectd.RegExPropertyExtender"> <2>
```
